### PR TITLE
WT-5441 Fix known visibility issues

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
This change is fixing a known issue with search (and one that I found while debugging).

- When embedding the location of an uncompressed value in a row, we should check that it is globally visible (i.e. that its commit timestamp is before oldest and there is no tombstone after it).
- There are a few places in search where we need to call `__wt_txn_read` to trigger a history store search. In this case, we were just searching the update list and then defaulting straight to the on-disk value.

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -331,6 +331,17 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_UPDATE **updp, bool *valid)
         cell = WT_COL_PTR(page, cip);
         if (__wt_cell_type(cell) == WT_CELL_DEL)
             return (0);
+        /* Check for an update. */
+        WT_RET(__wt_txn_read(session, cbt, cbt->ins == NULL ? NULL : cbt->ins->upd, &upd));
```
Comment (tetsuo-cpp) [inline src/btree/bt_cursor.c]: I think for the variable-length column store, we were only searching lookaside when we had an insert list. Instead of returning the on-disk value, let's call `__wt_txn_read` here and check all the usual suspects.

File: src/btree/bt_cursor.c [hunk 2]
```diff
@@ -350,17 +361,19 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_UPDATE **updp, bool *valid)
...
-            WT_RET(__wt_txn_read(session, cbt, page->modify->mod_row_update[cbt->slot], &upd));
-            if (upd != NULL) {
-                if (upd->type == WT_UPDATE_TOMBSTONE) {
-                    if (F_ISSET(upd, WT_UPDATE_RESTORED_FROM_DISK))
-                        __wt_free_update_list(session, &upd);
-                    return (0);
-                }
-                if (updp != NULL)
-                    *updp = upd;
+        WT_RET(__wt_txn_read(session, cbt,
```
Comment (tetsuo-cpp) [inline src/btree/bt_cursor.c]: Even a page with no modifications might need to lookaside search to find an update. In that case, just pass `NULL` and we'll go straight to checking on-disk and lookaside.

File: src/btree/bt_page.c [hunk 3]
```diff
@@ -570,7 +572,8 @@ __inmem_row_leaf(WT_SESSION_IMPL *session, WT_PAGE *page, bool check_unstable)
              * Simple keys without compression (not Huffman encoded or prefix compressed), can be
              * directly referenced on the page to avoid repeatedly unpacking their cells.
              */
-            if (!btree->huffman_key && unpack.prefix == 0)
+            if (!btree->huffman_key && unpack.prefix == 0 &&
+              (txn_global->has_oldest_timestamp && unpack.start_ts < txn_global->oldest_timestamp))
```
Comment (tetsuo-cpp) [inline src/btree/bt_page.c]: Only embed the location in the row for globally visible updates.

File: src/history/hs.c [hunk 4]
```diff
@@ -279,6 +279,12 @@ __wt_hs_cursor(WT_SESSION_IMPL *session, WT_CURSOR **cursorp, uint32_t *session_
                 if (!cache->hs_session_inuse[i]) {
                     *cursorp = cache->hs_session[i]->hs_cursor;
                     cache->hs_session_inuse[i] = true;
+                    if (reader) {
+                        cache->hs_session[i]->iface.begin_transaction(
```
Comment (tetsuo-cpp) [inline src/history/hs.c]: Since we're writing updates to the lookaside with timestamps, I think we need to set the read timestamp when finding lookaside updates. This explains why we're failing to find lookaside updates even in very obvious situations.

File: src/include/txn.i [hunk 5]
```diff
@@ -819,6 +840,10 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd, WT
 
     /* There is no BIRTHMARK in the history store file. */
     WT_ASSERT(session, upd == NULL || upd->type != WT_UPDATE_BIRTHMARK);
+
+    if (upd == NULL)
```
Comment (tetsuo-cpp) [inline src/include/txn.i]: Explicitly return tombstones when we get an absence of a value. This should stop the surrounding code from falling back to the on-disk value. That code should get ripped at some point but let's just aim for functional correctness first.

File: src/history/hs.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs.c]: Updates in history store has the same timestamp of the original update. It should be visible to user reads.

File: src/include/txn.i [hunk 6]
```diff
@@ -819,6 +840,10 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd, WT
```
Comment (quchenhao) [inline src/include/txn.i]: We can skip reading history store for in-memory database.

File: src/history/hs.c (hunk 4 above)
Comment (luke-pearson) [reply src/history/hs.c]: Isn't this to avoid reading a tombstone when you shouldn't?

File: test/suite/test_txn01.py [hunk 7]
```diff
@@ -101,6 +101,7 @@ def check(self, cursor, committed, total):
 
     # Loop through a set of inserts, periodically committing; before each
     # commit, verify the number of visible records matches the expected value.
+    @unittest.skip("Temporarily diasbled")
```
Comment (luke-pearson) [inline test/suite/test_txn01.py]: New failure then?

Comment (tetsuo-cpp) [reply test/suite/test_txn01.py]: Yes. That's why I've locked it.

File: src/history/hs.c (hunk 4 above)
Comment (tetsuo-cpp) [reply src/history/hs.c]: > Updates in history store have the same timestamp of the original updates. They should be visible to user reads.

But setting the timestamp is something you do to a session. And history store cursors have their own dedicated sessions. Therefore you have different visibility I think.

Does this seem right @kommiharibabu?

File: src/include/txn.i (hunk 6 above)
Comment (tetsuo-cpp) [reply src/include/txn.i]: That's what `WT_CONN_HS_OPEN` is checking.

File: src/btree/bt_page.c (hunk 3 above)
Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: Need to take into account stop timestamp as this gets lost when we do this.

File: src/history/hs.c (hunk 4 above)
Comment (tetsuo-cpp) [inline src/history/hs.c]: Borrow the isolation level from the current running transaction. If it's not snapshot then we can skip setting the read timestamp.

File: src/btree/bt_cursor.c [hunk 8]
```diff
@@ -331,6 +331,17 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_UPDATE **updp, bool *valid)
...
+        WT_RET(__wt_txn_read(session, cbt, cbt->ins == NULL ? NULL : cbt->ins->upd, &upd));
+        if (upd != NULL) {
+            if (upd->type == WT_UPDATE_TOMBSTONE) {
+                if (F_ISSET(upd, WT_UPDATE_RESTORED_FROM_DISK))
+                    __wt_free_update_list(session, &upd);
+                return (0);
+            }
+            if (updp != NULL)
+                *updp = upd;
+        }
```
Comment (keithbostic) [inline src/btree/bt_cursor.c]: There's a large comment at the start of this function, that looks like it needs updating? Not sure, to be clear, but maybe worth checking.

Comment (tetsuo-cpp) [reply src/btree/bt_cursor.c]: I'd rather defer this until a bit later since every function that calls `__wt_txn_read` is going to have undergo a significant refactor at some point. The comment has been wrong for a long time now since we began history store searching within `__wt_txn_read` in relevant history.

**Comment by tetsuo-cpp** [comment]
This is ready to go.

I'm a bit concerned with how `__wt_txn_read` is being used in general. We should probably refine this when we refactor that function. But for now, this change is getting us closer to being functionally correct so I'd like to get it in now.

File: src/btree/row_key.c [hunk 9]
```diff
@@ -315,13 +315,6 @@ __wt_row_leaf_key_work(
          * prefix compressed.
          */
         if (unpack->prefix == 0) {
-            /*
-             * The only reason to be here is a Huffman encoded key, a non-encoded key with no prefix
-             * compression should have been directly referenced, and we should not have needed to
-             * unpack its cell.
-             */
-            WT_ASSERT(session, btree->huffman_key != NULL);
```
Comment (quchenhao) [inline src/btree/row_key.c]: Why this assert is removed? Looks like irrelevant to the change.

File: src/reconcile/rec_row.c [hunk 10]
```diff
@@ -907,7 +907,7 @@ __wt_rec_row_leaf(
             kpack = &_kpack;
             __wt_cell_unpack(session, page, cell, kpack);
             if (btree->huffman_key == NULL && kpack->type == WT_CELL_KEY &&
-              tmpkey->size >= kpack->prefix) {
+              tmpkey->size >= kpack->prefix && tmpkey->size != 0) {
                 /*
                  * The previous clause checked for a prefix of zero, which means the temporary
```
Comment (quchenhao) [inline src/reconcile/rec_row.c]: If size cannot be zero, why not remove the assert below?

File: src/btree/row_key.c (hunk 9 above)
Comment (tetsuo-cpp) [reply src/btree/row_key.c]: This code path is hit when we don't encode the location into the row. Previously this only happens when we have a key that we're compressing with. Now we've added a stricter condition that takes into account the visibility of the unpacked value so this no longer holds.

File: src/reconcile/rec_row.c (hunk 10 above)
Comment (tetsuo-cpp) [reply src/reconcile/rec_row.c]: 635ba65

**Comment by tetsuo-cpp** [comment]
@keithbostic @luke-pearson 
Please take a look when you get the chance. I'd like to get this in today.

File: src/btree/bt_page.c [hunk 11]
```diff
@@ -570,7 +572,9 @@ __inmem_row_leaf(WT_SESSION_IMPL *session, WT_PAGE *page, bool check_unstable)
              * Simple keys without compression (not Huffman encoded or prefix compressed), can be
              * directly referenced on the page to avoid repeatedly unpacking their cells.
              */
-            if (!btree->huffman_key && unpack.prefix == 0)
+            if (!btree->huffman_key && unpack.prefix == 0 && unpack.stop_txn == WT_TXN_MAX &&
+              unpack.stop_ts == WT_TS_MAX && txn_global->has_oldest_timestamp &&
+              unpack.start_ts < txn_global->oldest_timestamp)
```
Comment (keithbostic) [inline src/btree/bt_page.c]: Wouldn't this be `<= oldest_timestamp`?

Comment (keithbostic) [inline src/btree/bt_page.c]: It might be worth updating the comment.

Comment (keithbostic) [inline src/btree/bt_page.c]: Can we ignore items where the `stop_ts < oldest_timestamp`?

Comment (keithbostic) [inline src/btree/bt_page.c]: I'm confused, keys don't have timestamps, do they?

File: src/include/txn.i [hunk 12]
```diff
@@ -746,6 +746,23 @@ __wt_txn_upd_visible(WT_SESSION_IMPL *session, WT_UPDATE *upd)
...
+{
+    size_t size;
+
+    WT_RET(__wt_update_alloc(session, NULL, updp, &size, WT_UPDATE_TOMBSTONE));
+    (*updp)->txnid = txnid;
+    (*updp)->durable_ts = (*updp)->start_ts = start_ts;
+    F_SET(*updp, WT_UPDATE_RESTORED_FROM_DISK);
+    return (0);
+}
+
```
Comment (keithbostic) [inline src/include/txn.i]: Reconciliation already has `__tombstone_update_alloc()`, and `__wt_find_hs_upd()`, `__wt_txn_read()`, `__rec_append_orig_value()`, `__wt_rec_upd_select()` all allocate a `WT_UPDATE` structure and immediately fill out the ts/txn information.

I've wondered for awhile if we need a variant of `__wt_update_alloc()` that takes timestamp/txn information, maybe that's a better way forward? It might help us avoid bugs by requiring we ensure all the ts/txn fields are filled in?

File: src/btree/bt_page.c (hunk 11 above)
Comment (bvpvamsikrishna) [inline src/btree/bt_page.c]: I think this comparison has to be "<=".

File: src/btree/bt_page.c [hunk 13]
```diff
@@ -587,7 +591,9 @@ __inmem_row_leaf(WT_SESSION_IMPL *session, WT_PAGE *page, bool check_unstable)
              * Simple values without compression can be directly referenced on the page to avoid
              * repeatedly unpacking their cells.
              */
-            if (!btree->huffman_value)
+            if (!btree->huffman_value && unpack.stop_txn == WT_TXN_MAX &&
+              unpack.stop_ts == WT_TS_MAX && txn_global->has_oldest_timestamp &&
+              unpack.start_ts < txn_global->oldest_timestamp)
```
Comment (bvpvamsikrishna) [inline src/btree/bt_page.c]: same as the previous comment about comparison operation.
In my mental model, I always imagine start to use "<=" and stop to use "<" as start timepair is inclusive of validity and stop is exclusive of validity. If for any reason this logic is not valid in this case, I think it deserves a comment.

File: src/btree/bt_cursor.c [hunk 14]
```diff
@@ -331,6 +331,23 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_UPDATE **updp, bool *valid)
...
+            if (upd != NULL) {
+                if (upd->type == WT_UPDATE_TOMBSTONE) {
+                    if (F_ISSET(upd, WT_UPDATE_RESTORED_FROM_DISK))
+                        __wt_free_update_list(session, &upd);
+                    return (0);
+                }
+                if (updp != NULL)
+                    *updp = upd;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply src/btree/bt_cursor.c]: Yeah, the `NULL` argument to `__wt_read_txn` is the in-memory update list we search (meaning we don't have one). That means we go straight to on-disk and then to history store. That's what the comment above was meant to indicate.

File: src/btree/bt_page.c (hunk 11 above)
Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: 417ce1c

File: src/btree/bt_page.c (hunk 13 above)
Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: Thanks! That's a good way to think of it.

File: src/btree/bt_page.c (hunk 11 above)
Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: 417ce1c

You're right. This check is redundant for key cells.

Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: 417ce1c

Comment (bvpvamsikrishna) [reply src/btree/bt_page.c]: I prefer not to ignore these when reading, if all entries are obsolete, then we will have an empty page ??
My preference is to ignore obsolete entries as part of reconciliation.

Comment (keithbostic) [reply src/btree/bt_page.c]: It's a performance cost on every search operation. I don't necessarily disagree with you, but it can cost us?

File: src/include/txn.i (hunk 12 above)
Comment (tetsuo-cpp) [reply src/include/txn.i]: Hi, I've made a ticket for the refactor so I think we can address it then: WT-5470.
Most of these usages should disappear.

**Comment by tetsuo-cpp** [comment]
> LGTM. Is there a follow up ticket for the refactor?

@luke-pearson Yep! WT-5470 is the ticket number.

Thanks for the reviews guys. Merging now...

File: src/btree/bt_page.c (hunk 11 above)
Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: @keithbostic Do we need to know ahead of time so we can skip the key cell also (as opposed to just the value)?

Comment (tetsuo-cpp) [reply src/btree/bt_page.c]: I've created WT-5471 to track this discussion.

Comment (keithbostic) [reply src/btree/bt_page.c]: We just back up in that case, there are examples in the function:
```
        case WT_CELL_VALUE_OVFL:
            /* Optionally skip unstable values */
            if (check_unstable && __unstable_skip(session, page->dsk, &unpack)) {
                --rip;
                --page->entries;
            }
```
